src/proof/nova.rs

```diff
@@ -1912,4 +1914,39 @@ mod tests {
             false,
         );
     }
+
+    #[test]
+    fn outer_prove_begin() {
+        let mut s = Store::<Fr>::default();
+
+        let src = "(begin (emit 1) (emit 2) (emit 3))";
+
+        let expr = s.read(src).unwrap();
+        let limit = 300;
+
+        let (
+            IO {
+                expr: _result_expr,
+                env: _new_env,
+                cont: _continuation,
+            },
+            _iterations,
+            emitted,
+        ) = Evaluator::new(expr, empty_sym_env(&s), &mut s, limit).eval();
```

Comment:
This isn't really checking that the circuit is correct. It's just tucking an evaluation test into what feels like it should be a circuit test.

One way to actually check the emitted values would be a series of calls to `outer_prove_aux` using the same input but different limits. That would allow you to stop the computation in the middle. Then you could check the output expr/continuation to demonstrate that the expected values were indeed emitted as intermediates of the final computation.

If the number of iterations used to reach each expected intermediate emitted values increases monotonically, then that demonstrates that they were emitted sequentially. There is probably a useful utility to be written to help extract the iteration-count to check for such tests. Likewise, there is probably a reasonable way of testing this which doesn't require re-running the whole computation for each intermediate checked. Rather, we could run it once while also checking that intermediate values were emitted either sequentially (in *some* set of constraint systems) or at specified intermediate values. The latter, being a little more explicit, may yield a better artifact for observing the behavior of real programs — but it might be a little more annoying to write without some utility/tooling support. Maybe the way to handle that is to have the enhanced test helper take an optional list of intermediate values to verify. That would make it easy to instrument the tests to harvest the real values, which could then be added to the test itself, making the information available later by inspecting the test as written.